fendermint/vm/interpreter/src/lib.rs

```diff
@@ -0,0 +1,79 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+use async_trait::async_trait;
+
+pub mod bytes;
+pub mod chain;
+mod externs;
+pub mod fvm;
+pub mod signed;
+
+/// Unix timestamp (in seconds) of the current block.
+pub struct Timestamp(pub u64);
+
+/// The `Interpreter`applies messages on some state, which is
+/// tied to the lifecycle of a block in the ABCI.
+///
+/// By making it generic, the intention is that interpreters can
+/// be stacked, changing the type of message along the way. For
+/// example on the outermost layer the input message can be a mix
+/// of self-contained messages and CIDs proposed for resolution
+/// or execution, while in the innermost layer it's all self-contained.
+/// Some interpreters would act like middlewares to resolve CIDs into
```

Comment:
I love this, this is going to be key, we should consider using this interpreter abstraction also as part of the IPC agent to get this translation.

Reply:
I went back and forth a lot on this. It's not clear where to put state exactly: into the middleware, or into the `State` that goes in and out. I wanted to change it to return effects explicitly, and otherwise be synchronous, but then I realised that I will have to resolve CIDs on the fly anyway, when they are given to me for execution. So, for now, I'll see where it goes.